Add neopixel support for up to 85 LEDs in a single chain

KevinOConnor opened the pull request:
Increase maximum neopixel chain length from 18 to 85.

I don't have a good way to test this on long LED chains, so please report success (or failure) here.

`cd ~/klipper ; git fetch ; git checkout origin/work-neopixel-20200928 ; sudo service klipper stop ; make ; make flash ; sudo service klipper start`

-Kevin

mental405 commented:
logging.info() is called in neopixel.py but the logging library is not being imported. 

It appears to be working as intended.
https://youtu.be/iML3OV9HTdM

There are a lot of errors coming through in the log 

``Neopixel update did not succeed``

This is likely due to my shoehorning frame data through without giving it a timer.

I wrote a quick gcode macro to set random colors along the length of the strip using SET_LED and it is not throwing errors.   

https://www.youtube.com/watch?v=JDYKMuvsvFw

KevinOConnor commented:
Thanks.  I fixed the logging import.

I guess you are reporting that the code on this branch works, but it is not compatible with custom changes on your branch?  If so, that makes sense.  The new code on this branch would need a mutex if the  `send_data()` method were called from anything other than a gcode command.  (To ensure an update is "atomic".)

-Kevin

mental405 commented:
I am not super clear on how mutex's work in klipper and python but if attaching one to either send_data() or having the currently executing send_data() lock the color_data list of the neopixel object until sending is complete, that would probably solve it.

altvnk commented:
I have 48 NeoPixels LED strip for my case light. So I switched to this PR right when it was introduced, it was very annoying to execute separate Python script to set up lights. Now everything is in my macro's and works great. Looking forward for having this in the main branch. 
Thanks!

KevinOConnor commented:
FYI, I updated this branch to add experimental support for RGBW LEDs.  I also removed the limit of 85 LEDs (there's now just an arbitrary sanity limit of 500 bytes per chain - which is 166 standard LEDs or 125 RGBW LEDs).

-Kevin

al3ph commented:
[neopixel pixel]
pin = PA8
chain_count = 20
color_order = RGBW
initial_red = 0.0
initial_green = 0.0
initial_blue = 0.0
initial_white = 0.0

[virtual_sdcard]
path = ~/sdcard

[pause_resume]

[display_status]

[stepper_x]
step_pin = PB13
dir_pin = !PB12
enable_pin = !PB14
step_distance = .0125
endstop_pin = PC0
position_endstop = 120
position_max = 120
homing_speed = 20
homing_retract_dist = 5
homing_positive_dir = true

[tmc2209 stepper_x]
uart_pin = PC11
tx_pin = PC10
uart_address = 0
microsteps = 16
interpolate = True
run_current = .25
hold_current = .25
sense_resistor = 0.110
stealthchop_threshold = 500

[stepper_y]
step_pin = PB10
dir_pin = !PB2
enable_pin = !PB11
step_distance = .0125
endstop_pin = PC1
position_endstop = 120
position_max = 120
homing_speed = 20
homing_retract_dist = 5
homing_positive_dir = true

[tmc2209 stepper_y]
uart_pin = PC11
tx_pin = PC10
uart_address = 2
microsteps = 16
interpolate = True
run_current = .25
hold_current = .25
sense_resistor = 0.110
stealthchop_threshold = 500

[stepper_z]
step_pin = PB0
dir_pin = !PC5
enable_pin = !PB1
step_distance = .0025
endstop_pin = PC2
position_endstop = -.10
position_max = 120
position_min = -1
homing_speed = 10
second_homing_speed = 3.0
homing_retract_dist = 3.0

[tmc2209 stepper_z]
uart_pin = PC11
tx_pin = PC10
uart_address = 1
microsteps = 16
interpolate = True
run_current = .25
hold_current = 0.25
sense_resistor = 0.110
stealthchop_threshold = 500

[extruder]
step_pin = PB3
dir_pin = !PB4
enable_pin = !PD2
step_distance = .0024
nozzle_diameter = 0.400
filament_diameter = 1.750
heater_pin = PC8
sensor_type = NTC 100K beta 3950
sensor_pin = PA0
min_temp = 0
max_temp = 270
min_extrude_temp = 170
max_extrude_only_distance = 780.0
max_extrude_cross_section = .8
pressure_advance = 0.0
pressure_advance_smooth_time = 0.040
control = pid
pid_kp = 21.238
pid_ki = 1.388
pid_kd = 81.236

[tmc2209 extruder]
uart_pin = PC11
tx_pin = PC10
uart_address = 3
microsteps = 16
interpolate = True
run_current = 0.2
hold_current = 0.2
sense_resistor = 0.110
stealthchop_threshold = 500

[heater_bed]
heater_pin = !PC15
sensor_type = NTC 100K beta 3950
sensor_pin = PC3
smooth_time = 3.0
max_power = 1.0
control = pid
min_temp = 0
max_temp = 120
pid_kp = 58.437
pid_ki = 2.347
pid_kd = 363.769

[printer]
kinematics = corexy
max_velocity = 250
max_accel = 2000
max_z_velocity = 10
max_z_accel = 30
square_corner_velocity = 5.0

[heater_fan hotend_fan]
pin = PC7
max_power = 1.0
kick_start_time = 0.5
heater = extruder
heater_temp = 50.0

[fan]
pin = PC6
kick_start_time = 0.5
off_below = 0.5
cycle_time = 0.001

[idle_timeout]
timeout = 1800

[homing_override]
axes = z
set_position_z = 0
gcode = 
	G90
	G0 Z5 F600
	G28 X Y
	G0 X30 Y0 F3600
	G28 Z
	G0 Z10 F1800
	G0 X60 Y60 Z20 F3600

[bed_screws]
screw1 = 60,5
screw1_name = front screw
screw2 = 5,115
screw2_name = back left
screw3 = 115,115
screw3_name = back right

[gcode_macro PRINT_START]
gcode = 
	G28                            ; home all axes
	G1 Z20 F3000                   ; move nozzle away from bed

[gcode_macro PRINT_END]
gcode = 
	M400                           ; wait for buffer to clear
	G92 E0                         ; zero the extruder
	G1 E-4.0 F3600                 ; retract filament
	G91                            ; relative positioning
	G0 Z1.00 X20.0 Y20.0 F20000    ; move nozzle to remove stringing
	TURN_OFF_HEATERS
	M107                           ; turn off fan
	G1 Z2 F3000                    ; move nozzle up 2mm
	G90                            ; absolute positioning
	G0  X60 Y120 F3600            ; park nozzle at rear

[gcode_macro LOAD_FILAMENT]
gcode = 
	M83                            ; set extruder to relative
	G1 E365 F1800                  ; quickly load filament to down bowden
	G1 E30 F300                    ; slower extrusion for hotend path
	G1 E15 F150                    ; prime nozzle with filament
	M82                            ; set extruder to absolute

[gcode_macro UNLOAD_FILAMENT]
gcode = 
	M83                            ; set extruder to relative
	G1 E10 F300                    ; extrude a little to soften tip
	G1 E-440 F1800                 ; retract filament completely
	M82                            ; set extruder to absolute

[gcode_macro global]
variable_led_value = 0.0
gcode = 
	SET_GCODE_VARIABLE MACRO=global VARIABLE=led_value VALUE=0.0

[gcode_macro LED_OFF]
gcode = 
	SET_GCODE_VARIABLE MACRO=global VARIABLE=led_value VALUE=0.0
	SET_LED LED=pixel RED=0.0 GREEN=0.0 BLUE=0.0

[gcode_macro LED_ON]
gcode = 
	SET_GCODE_VARIABLE MACRO=global VARIABLE=led_value VALUE={[printer["gcode_macro global"].led_value*2.0,0.03125]|max}
	SET_LED LED=pixel RED={[printer["gcode_macro global"].led_value*2.0,0.03125]|max} GREEN={[printer["gcode_macro global"].led_value*2.0,0.03125]|max} BLUE={[printer["gcode_macro global"].led_value*2.0,0.03125]|max}

[gcode_macro ZUP]
gcode = 
	SET_GCODE_OFFSET Z_ADJUST=0.05 MOVE=1

[gcode_macro ZDOWN]
gcode = 
	SET_GCODE_OFFSET Z_ADJUST=-0.05 MOVE=1

[mcu]
serial = /dev/ttyAMA0
baud = 250000
restart_method = command
=======================
Unhandled exception during connect
Traceback (most recent call last):
  File "/home/pi/klipper/klippy/klippy.py", line 146, in _connect
    self._read_config()
  File "/home/pi/klipper/klippy/klippy.py", line 139, in _read_config
    self.load_object(config, section_config.get_name(), None)
  File "/home/pi/klipper/klippy/klippy.py", line 128, in load_object
    self.objects[section] = init_func(config.getsection(section))
  File "/home/pi/klipper/klippy/extras/neopixel.py", line 119, in load_config_prefix
    return PrinterNeoPixel(config)
  File "/home/pi/klipper/klippy/extras/neopixel.py", line 40, in __init__
    self.update_color_data(red, green, blue, white)
  File "/home/pi/klipper/klippy/extras/neopixel.py", line 70, in update_color_data
    self.color_data[:] = color_data * self.chain_count
TypeError: an integer or string of size 1 is required

al3ph commented on klippy/extras/neopixel.py:
```diff
@@ -21,27 +22,33 @@ def __init__(self, config):
+                                         minval=1, maxval=255//3)
+        self.neopixel_update_cmd = self.neopixel_send_cmd = None
         # Initial color
+        self.color_data = bytearray(self.chain_count * 3)
```
Sure about this ?

al3ph replied:
self.chain_count * 4 ?

al3ph replied:
Ignore me looking at the wrong commit doh..

al3ph commented on klippy/extras/neopixel.py:
```diff
@@ -42,25 +47,28 @@ def build_config(self):
+    def update_color_data(self, red, green, blue, white, index=None):
         red = int(red * 255. + .5)
         blue = int(blue * 255. + .5)
         green = int(green * 255. + .5)
```
Missing white = int(white *255.+.5)

al3ph replied:
This removes the python error, but I think it's crashing my controller, I'm gonna reboot just to be sure..

al3ph replied:
Working now, but the pixel is not RGBW it's GRBW :)

KevinOConnor commented:
Okay, thanks.  I updated the code on this branch (including the addition of GRBW).  Hopefully it works now.

-Kevin

al3ph commented:
Working great! Thanks!!

mental405 commented:
I don't know how far down the color-reproduction-visual-perception rabbit hole anyone is willing to go, but RBG and RGBW have different color spaces but they can be converted with a little math and using an HSB/HSV/HSL color space instead of RGB.

I think the basic premise is RGB is first converted to a Hue, Saturation, and Brightness, then the W is derived from the Saturation value and then that is applied to the brightness... or something.

It might make the gcode commands more agnostic if they all only took R G and B parameters.

KevinOConnor commented:
> I don't know how far down the color-reproduction-visual-perception rabbit hole anyone is willing to go

FWIW, that's not something I planning on working on.  But it's fine if someone else wishes to implement it.  I'd guess it would still be useful to be able to directly control the individual LED components even with hue/saturation/brightness support though.

-Kevin

al3ph commented:
I'd not touch that with a barge pole.. :) It's a huge can of worms. I 
work in the film industry, so we deal with ACES colour workflows, so I 
know how messy this stuff gets. For a start RGBW light come in three 
varieties with different white temperatures. I imagine getting vaguely 
accurate colours is hard enough without mixing white :)

HSV is handy for doing rainbows though.

On 11/10/2020 16:21, KevinOConnor wrote:
>     I don't know how far down the color-reproduction-visual-perception
>     rabbit hole anyone is willing to go
> 
> FWIW, that's not something I planning on working on. But it's fine if 
> someone else wishes to implement it. I'd guess it would still be useful 
> to be able to directly control the individual LED components even with 
> hue/saturation/brightness support though.
> 
> -Kevin
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub 
> <https://github.com/KevinOConnor/klipper/pull/3380#issuecomment-706720808>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AAH2NXG6CE7JZC6WZ4E5F2DSKHEP3ANCNFSM4R425OTQ>.
>

al3ph commented:
I've noticed when I set the colour during a print the print locks up for almost a second. Is that intentional ?

KevinOConnor commented:
I'm not sure what you mean by "locks up".  The SET_LED command will fully flush the lookahead queue, which will add a pause between queued moves - if it is issued during a print.

It is possible to avoid flushing the lookahead queue, but it is more code work to do it that way.

-Kevin

al3ph commented:
The print pauses, just checking this is expected, I imagined that setting the LED wouldn't effect the print (i.e. it'd be milli seconds)

KevinOConnor commented:
FYI, I updated this branch so that it will no longer flush the lookahead queue on a SET_LED command.

-Kevin

al3ph commented:
Cool, I assume this requires reflashing the firmware ?
